hazelcast-client/src/main/java/com/hazelcast/client/map/impl/nearcache/ClientHeapNearCache.java

```diff
@@ -260,4 +196,71 @@ public NearCacheStatsImpl getNearCacheStats() {
         stats.setOwnedEntryMemoryCost(ownedEntryMemoryCost);
         return stats;
     }
+
+    private void fireCacheEviction() {
+        if (canEvict.compareAndSet(true, false)) {
+            try {
+                executionService.executeInternal(new Runnable() {
+                    public void run() {
+                        try {
+                            Set<NearCacheRecord> records = new TreeSet<NearCacheRecord>(selectedComparator);
+                            records.addAll(cache.values());
+                            int evictSize = (int) (cache.size() * EVICTION_FACTOR);
+                            int i = 0;
+                            for (NearCacheRecord record : records) {
+                                cache.remove(record.getKey());
+                                stats.incrementEvictions();
+                                if (++i > evictSize) {
+                                    break;
+                                }
+                            }
+                        } finally {
+                            canEvict.set(true);
+                        }
+                        if (cache.size() >= maxSize && canEvict.compareAndSet(true, false)) {
+                            try {
+                                executionService.executeInternal(this);
+                            } catch (RejectedExecutionException e) {
+                                canEvict.set(true);
+                            }
+                        }
+                    }
+                });
+            } catch (RejectedExecutionException e) {
+                canEvict.set(true);
+            } catch (Exception e) {
+                throw rethrow(e);
```

Comment:
Are you not at risk of not releasing the 'canEvict' flag if another exception is thrown?

Perhaps have a single catch which always restores the canEvict flag?